server/service/src/programs/contact_trace/contact_trace_updated.rs

```diff
@@ -0,0 +1,90 @@
+use super::{
+    contact_trace_schema::{SchemaContactTrace, SchemaGender},
+    upsert::UpsertContactTraceError,
+};
+use chrono::{DateTime, NaiveDate};
+use repository::{
+    contact_trace::{ContactTraceFilter, ContactTraceRepository},
+    contact_trace_row::{ContactTraceRow, ContactTraceRowRepository},
+    Document, Gender, ProgramRow, StorageConnection, StringFilter,
+};
+use std::str::FromStr;
+use util::hash::sha256;
+
+/// Callback called when a program enrolment document has been updated
+pub(crate) fn update_contact_trace_row(
+    con: &StorageConnection,
+    patient_id: &str,
+    document: &Document,
+    contact_trace: SchemaContactTrace,
+    program_row: ProgramRow,
+) -> Result<ContactTraceRow, UpsertContactTraceError> {
+    let datetime = DateTime::parse_from_rfc3339(&contact_trace.datetime)
+        .map_err(|err| {
+            UpsertContactTraceError::InternalError(format!(
+                "Invalid enrolment datetime format: {}",
+                err
+            ))
+        })?
+        .naive_utc();
+
+    // Retrieve existing or create now contact trace id
+    let repo = ContactTraceRepository::new(con);
+    let contact_trace_row = repo
+        .query_by_filter(ContactTraceFilter {
+            document_name: Some(StringFilter::equal_to(&document.name)),
+            ..ContactTraceFilter::default()
+        })?
+        .pop();
+    let contact_trace_id = match contact_trace_row {
+        Some(contact_trace_row) => contact_trace_row.0.id,
+        None => sha256(&document.name),
+    };
+
+    let store_id = contact_trace
```

Comment:
This method is also used when integrating the contact trace during sync and where you only have the document. 

This store_id is supposed to be the store_id where the encounter happened, the idea was that you could also record a location that is not a store, e.g. at somebody's house as recorded on a paper form. However, I wonder if an encounter always has to be associated with a real store_id. For example, even now we always use the store from where the encounter has been imported. We use this store_id for reporting as well.

@richardmoizeau think I will change the schema so that an encounter has to be associated with a mSupply store.

Reply:
> @richardmoizeau think I will change the schema so that an encounter has to be associated with a mSupply store.

This would help in reporting too right ? And apart from reporting it shouldn't have any other effect right (and store_id will be added to document ?)

Reply:
> (and store_id will be added to document ?)

In general a document doesn't belong to a particular store. For example, patients or program enrolments. IMHO it makes sense to associate a contact trace and an encounter with a store. With program enrolments I find it a bit more tricky, but I would say its independent of the store...